File: src/common/web/core/messaging/composers/CommandReplyComposer.ts
import { UnitID } from "../../../utils/UnitID";
import { Channel } from "../Channel";
import { Command } from "../Command";
import { CommandReply } from "../CommandReply";
import { type ConstructableMessage } from "../Message";
import { type MessageBusProtocol } from "../MessageBusProtocol";
import { CommandReplyMetaInformation } from "../meta/CommandReplyMetaInformation";
import { MessageEntrypoint, MessageMetaInformation } from "../meta/MessageMetaInformation";
import { MessageComposer } from "./MessageComposer";

/**
 * Composer for ``CommandReply`` messages.
 */
export class CommandReplyComposer<MsgType extends CommandReply> extends MessageComposer<MsgType> {
    private readonly _command: Command;

    public constructor(originID: UnitID, messageBus: MessageBusProtocol, msgType: ConstructableMessage<MsgType>, cmd: Command,
                       params: Record<string, any> = {}) {
        super(originID, messageBus, msgType, params, cmd);

        this._command = cmd;
    }

    /**
     * Sends the built message through the message bus.
     */
    public emit(): void {
        let target = this._command.origin.equals(this._originID) ? Channel.local() : Channel.direct(this._command.origin.toString());
        super.emit(target);
    }

    protected createMetaInformation(): MessageMetaInformation {
        return new CommandReplyMetaInformation(MessageEntrypoint.Local);
    }
}

File: src/common/py/core/messaging/composers/command_reply_composer.py
from .message_composer import MessageComposer
from .. import MessageType, MessageBusProtocol, CommandType, Channel
from ..meta import MessageMetaInformation
from ....utils import UnitID


class CommandReplyComposer(MessageComposer):
    """
    Composer for ``CommandReply`` messages.
    """

    def __init__(
        self,
        origin_id: UnitID,
        message_bus: MessageBusProtocol,
        msg_type: type[MessageType],
        command: CommandType,
        **kwargs,
    ):
        """
        Args:
            origin_id: The component identifier of the origin of newly created messages.
            message_bus: The global message bus to use.
            msg_type: The message type.
            command: The ``Command`` this reply is based on.
            **kwargs: Additional message parameters.
        """
        super().__init__(origin_id, message_bus, msg_type, command, **kwargs)

        self._command = command

    def emit(self) -> None:  # pylint: disable=arguments-differ
        """
        Sends the built message through the message bus.
        """
        target = (
            Channel.local()
            if self._command.origin.equals(self._origin_id)
            else Channel.direct(self._command.origin)
        )
        super().emit(target)

    def _create_meta_information(self) -> MessageMetaInformation:
        from ..meta import CommandReplyMetaInformation

        return CommandReplyMetaInformation(
            entrypoint=MessageMetaInformation.Entrypoint.LOCAL
        )
